xdp-document.c

```diff
@@ -181,6 +181,40 @@ xdp_document_get_permissions (XdpDocument *doc,
   return flags;
 }
 
+static void
+permissions_saved (GObject *source_object,
+                   GAsyncResult *result,
+                   gpointer data)
+{
+  GomResource *resource = GOM_RESOURCE (source_object);
+  GError *error = NULL;
+
+  if (!gom_resource_save_finish (resource, result, &error))
+    {
+      g_warning ("Failed to save permissions: %s", error->message);
+      g_error_free (error);
+    }
+}
+
+void
+xdp_document_grant_permissions (XdpDocument *doc,
+                                const char *app_id,
+                                XdpPermissionFlags perms)
+{
+  XdpPermissions *permissions = NULL;
+  XdpPermissionFlags old_perms;
+  GomRepository *repo;
+
+  old_perms = xdp_document_get_permissions (doc, app_id);
+  if ((perms & old_perms) == perms)
+    return;
+
+  g_object_get (doc, "repository", &repo, NULL);
+  permissions = xdp_permissions_new (repo, doc, app_id, perms, FALSE);
+  doc->permissions = g_list_prepend (doc->permissions, permissions);
```

Comment:
It seems more right to not prepend the new permissions object until the save is finished, no?